R/transition_risk_profile.R

```diff
@@ -0,0 +1,118 @@
+#'
+#' restore <- options(readr.show_col_types = FALSE)
+#'
+#' products <- read_csv(toy_emissions_profile_products_ecoinvent())
+#' # TODO: Submit a PR to tiltToyData?
+#' products_scenarios <- read_csv(toy_sector_profile_companies()) |>
+#'   select(-c("companies_id", "company_name", "clustered")) |>
+#'   distinct()
+#' scenarios <- read_csv(toy_sector_profile_any_scenarios())
+#'
+#' transition_risk_profile(products, products_scenarios, scenarios)
+#'
+#' options(restore)
+transition_risk_profile <- function(products,
+                                    # TODO: Do we need a new dataset in tiltToyData?
+                                    products_scenarios,
+                                    scenarios,
+                                    low_threshold = tercile(.data$transition_risk_score, 1),
+                                    high_threshold = tercile(.data$transition_risk_score, 2)) {
+  score_transition_risk(products, products_scenarios, scenarios) |>
+    # Add thresholds
+    mutate(
+      transition_risk_low_threshold = !!rlang::enquo(low_threshold),
+      transition_risk_high_threshold = !!rlang::enquo(high_threshold),
+      .by = "benchmark_tr_score"
+    ) |>
+    # ... do something else?
+    # Polish
+    select(-c("scenario", "year"))
+}
+
+tercile <- function(x, which) {
+  out <- quantile(x, probs = c(1 / 3, 2 / 3), na.rm = TRUE)
+  out[[which]]
+}
+
+score_transition_risk <- function(products,
+                                  products_scenarios,
+                                  scenarios) {
+  ranking <- epa_compute_profile_ranking(products)
+
+  target <- spa_compute_profile_ranking(
+    products_scenarios,
+    scenarios
+  )
+
+  ranking_target <- full_join(
+    ranking |> select_crucial_ranking(),
+    target |> select_crucial_target(),
+    by = c("activity_uuid_product_uuid"),
+    relationship = "many-to-many"
+  )
+
+  ranking_target |>
+    add_transition_risk_score("profile_ranking", "reductions") |>
+    add_benchmark_tr_score() |>
+    set_na_benchmark_tr_score() |>
+    distinct()
+}
+
+select_crucial_target <- function(data) {
+  data |>
+    select(c("activity_uuid_product_uuid", "scenario", "year", "reductions")) |>
+    # These two steps seem to belong closer to the edge
+    mutate(scenario_year = paste(.data$scenario, .data$year, sep = "_"))
+}
+
+select_crucial_ranking <- function(data) {
+  data |>
+    select(c("activity_uuid_product_uuid", "grouped_by", "profile_ranking"))
+}
+
+add_transition_risk_score <- function(data,
+                                      col_ranking = "profile_ranking",
+                                      col_target = "reductions") {
+  transition_risk <- function(x, target) {
+    ifelse(is.na(x) | is.na(target), NA, (x + target) / 2)
+  }
+
+  x <- data[[col_ranking]]
+  target <- data[[col_target]]
+  mutate(data, transition_risk_score = transition_risk(x, target))
+}
+
+add_benchmark_tr_score <- function(data) {
+  data |>
```

Comment:
@maurolepore I can see that you have separated them into two functions (which is also right). However, by using a single function, my aim is to export the `create_tr_benchmarks_tr_score` from `tiltIndicator/tiltTransitionRisk` and use it in tiltIndicatorAfter, which is already present in in tiltIndicatorAfter (please check [here](https://github.com/2DegreesInvesting/tiltIndicatorAfter/blob/6b80b80459edbb47750b88f4c794aa7a26d925dc/R/score_transition_risk.R#L95)). Should we now export two functions to tiltIndicatorAfter?

Reply:
In general, the smaller and the more general the functions are, the more reusable they become. 

The component tiltTransitionRisk can stay as a dev-oriented, high-level  component -- with general, small functions that know nothing about column names -- just core business logic.

Because it's dev-oriented and new, we can conveniently export everything and re-import it from titlIndicatorAfter. There you can wrap multiple small, general functions into the bigger, more detailed functions that you want.

This silly example captures that idea, of hosting small general code in one high-level package and wrapping it from a lower-level package.

```r
# tiltBusinessLogic package (high-level)

#' @export
pick_col <- function(data, col)  {
  select(data, all_of(col))
}

#' @export
pick_row <- function(data, ...) {
  filter(data, ...)
}


# tiltDetails package (low-level)

subset_benchmark <- function(data, .benchmark) {
  stopifnot(hasName(data, "companies_id")
  stopifnot(hasName(data, "benchmark")
  
  tiltBusinessLogic::pick_col(data, "companies_id") |>
    tiltBusinessLogic::pick_row(benchmark == .benchmark)
}

Reply:
Note from tiltIndicatorAfter you can import everything in tiltTransitionRisk with

```r
#' @import tiltTransitionRisk
NULL
```

This is generally considered bad practice but in this case doing this temporarily I think it's fine -- the separation across components is mostly a convenience for development, and what we import is code owned by ourselves.

Reply:
@maurolepore I didn't get you properly here. Do you mean I can use these lines of code in tiltIndicatorAfter if I want to use `tiltTransitionRisk`. right?